### DISCUSSION TITLE
Optional Query Parameters Don't Generate Correct OpenAPI Schema

### DISCUSSION THREAD

**Discussion by pythonweb2** [fastapi/fastapi#10481] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
from fastapi import Query

app = FastAPI()


@app.get("/works")
async def works(x: int = Query(default=None)):
    return {"test": x}


@app.get("/does_not_work")
async def does_not_work(x: int | None):
    return {"test": x}
```

### Description

* Open the browser and go to the docs page
* The "works" endpoint has OpenAPI schema that lets the user know that the query parameter is an integer.
* The "does_not_work" endpoint does not

![image](https://github.com/tiangolo/fastapi/assets/32141163/c014a236-ce6f-418d-8579-9a2d0973cc5e)

![image](https://github.com/tiangolo/fastapi/assets/32141163/d72b9dd4-6f38-4e4a-8655-56576d7bf59f)

### Operating System

Windows

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

3.10.10 (Also was seeing this on 3.11)

**Comment by pythonweb2** [comment]
I tried this with `pydantic==1.10.13`, and it was slightly better (it got the types correct), but in the second example it marked the query parameter as required.

```
> poetry show
anyio             3.7.1   High level compatibility layer for multiple asynchronous event loop implementations
click             8.1.7   Composable command line interface toolkit
colorama          0.4.6   Cross-platform colored terminal text.
fastapi           0.104.0 FastAPI framework, high performance, easy to learn, fast to code, ready for production
h11               0.14.0  A pure-Python, bring-your-own-I/O implementation of HTTP/1.1
idna              3.4     Internationalized Domain Names in Applications (IDNA)
pydantic          1.10.13 Data validation and settings management using python type hints
sniffio           1.3.0   Sniff out which async library your code is running under
starlette         0.27.0  The little ASGI library that shines.
typing-extensions 4.8.0   Backported and Experimental Type Hints for Python 3.8+
uvicorn           0.23.2  The lightning-fast ASGI server.
```

![image](https://github.com/tiangolo/fastapi/assets/32141163/24e858c8-58f5-4d04-945f-7bf8a23ddb0a)

![image](https://github.com/tiangolo/fastapi/assets/32141163/7cf0168d-b8b3-4be3-9425-e90c4522ef18)

**Reply by pythonweb2** [reply]
Based on the comments from @Ale-Cas it seems like it actually does work totally fine with Pydantic v1, so perhaps this is a pydantic v2 issue?

**Comment by Ale-Cas** [comment]
Hi @pythonweb2 in order to avoid the `x` parameter being required you just need to use None as the default value in the second endpoint, this should solve your problem:
```python
@app.get("/does_not_work")
async def does_not_work(x: int | None = None):
    return {"test": x}
```
<img width="615" alt="Screenshot 2023-10-20 at 11 25 25" src="https://github.com/tiangolo/fastapi/assets/64859146/7e68228c-e69f-4578-ac45-922ffc49732b">

**Reply by pythonweb2** [reply]
The openapi schema is still wrong though, notice how it doesn't indicate that the field is an integer type.

**Comment by Ale-Cas** [comment]
I'm also getting the validation error as expected when trying to pass a non-integer value
<img width="697" alt="Screenshot 2023-10-20 at 11 28 29" src="https://github.com/tiangolo/fastapi/assets/64859146/896d5b3c-619a-48f9-9324-3383c755ec58">

**Reply by pythonweb2** [reply]
The openapi schema is still wrong though, notice how it doesn't indicate that the field is an integer type.

**Comment by pythonweb2** [comment, marked as answer]
The fix seems to be in progress here: #9873
